### PR TITLE
RSDK-6963: Pull gRPC handling of video streams from gostream up to a higher level package.

### PR DESCRIPTION
**Pull request by dgottlieb**
This is 95% a code move. There's a little code deletion/refactor that (I think) greatly simplifies the type hierarchy. I'll note that inline with the diffs.

### CODE REVIEW

File: gostream/standalone_stream_server.go
Comment (dgottlieb) [inline gostream/standalone_stream_server.go:3]: This entire file was really just an example server that got pulled in with the gostream migration into the RDK. There's no dependency on this from any of RDK/goutils/app. I asked around a couple weeks ago if something else is magically dependent on this and got nothing.

File: robot/web/stream/server.go
Comment (dgottlieb) [inline robot/web/stream/server.go:19]: As per Go's conventions, I changed `StreamServer` -> `Server`. Because "Stream" is already a part of the package name.

More importantly, I got rid of the interface and just returned the bare structure type.

The only consequence of removing the interface was that the "standalone server" relied on it. Thus me just deleting that file.

File: robot/web/stream/server.go
Comment (dgottlieb) [inline robot/web/stream/server.go:20]: Before there was the `streamServer` and `streamRPCServer`. The latter technically implemented all of the `ListStreams`/`AddStream`/`RemoveStream` gRPC endpoints. I didn't see a need to keep them separate.

The consequence of that merging of types was this `Server` now needs to embed the "unimplemented stream service server" structure.

File: robot/web/stream/server.go
Comment (dgottlieb) [inline robot/web/stream/server.go:63]: One of the bits that annoyed me about the two "stream server" types was that we had the `streamServer.AddStream` call and the `streamRPCServer.AddStream` call. And I kept wanting the latter, but searching through text kept pointing me here first.

File: robot/web/stream/server.go
Comment (dgottlieb) [inline robot/web/stream/server.go:238]: Any changes to the relative position of where the code lives in this file were completely incidental. Happy to move things around or even split this type into its own file.

File: rimage/cmd/stream_camera/main.go
Comment (dgottlieb) [inline rimage/cmd/stream_camera/main.go:1]: I got the OK from Eliot to delete this file.

File: robot/web/stream/server.go
Comment (nicksanford) [inline robot/web/stream/server.go:111]:
```suggestion
	ss.mu.RLock()
	defer ss.mu.RUnlock()
```

File: robot/web/stream/server.go
Comment (nicksanford) [inline robot/web/stream/server.go:116]:
```suggestion
```

File: robot/web/stream/server.go
Comment (nicksanford) [inline robot/web/stream/server.go:128]: I see that `AddStream` doesn't lock `mu` before accessing `ss.streams`. Is `streams` not protected by `mu`? if not, could we please move all `Server` struct members protected by `mu` below `mu` & everything not protected by it above `mu`?

File: robot/web/stream/server.go
Comment (nicksanford) [inline robot/web/stream/server.go:158]: I know this is how the original code worked, so just for my own edification:

Do we really want to hold this lock while `ps.stream.Stop()` called, or are we just trying to hold it while we mutate `ps.stream.Stop()`?

File: robot/web/stream/server.go
Comment (nicksanford) [inline robot/web/stream/server.go:69]: This is mostly for my edification, I don't think how this is will hurt anything:
Why do we need to lock `mu` here? 

No other thread has access to `ss` outside of this function (as we just created it). Is there really any risk of a data race if we write to streams w/o a mutex before returning the `ss` pointer?

File: robot/web/stream/server.go
Comment (nicksanford) [inline robot/web/stream/server.go:223]: Same question regarding whether or not we need to lock `mu` before accessing `ss.streams`

File: robot/web/stream/server.go
Comment (nicksanford) [inline robot/web/stream/server.go:262]: Aren't we writing to `ss.streams[N].activePeers`? If so, why are we taking a read lock rather than a write lock?

File: robot/web/stream/server.go
Comment (nicksanford) [reply robot/web/stream/server.go:52]: never mind. I now understand that this is so we don't fail to compile if new methods are added to the StreamServerInterface which are not immediately implemented by `Server`

File: robot/web/stream/server.go
Comment (dgottlieb) [reply robot/web/stream/server.go:126]: This is a bit of a mess. Now that you've highlighted this, I need to restrain myself from just doing all of this over.

Notably there's a [write-write race](https://github.com/viamrobotics/rdk/pull/3699/files#diff-b664e7f5f24e04be1ed075bf7d0f4209eaf9fc88f34eb7acc987ad1693857223R266) if multiple threads call `Server.Stop` at the same time.

File: robot/web/stream/server.go
Comment (dgottlieb) [reply robot/web/stream/server.go:156]: It's unclear to me what the [concurrency contract](https://github.com/viamrobotics/rdk/blob/c629c2d3fc963777c719d0af3d6aa216929a482d/gostream/stream.go#L44-L45) is of `Stream.Stop`.

You know the implementing side of the interface a bit better than I do at the moment. Given the number of concurrency related smells, I'm inclined to leave this as-is for now. But I won't be surprised if you start making some drive-by changes in your PRs.

File: robot/web/stream/server.go
Comment (dgottlieb) [reply robot/web/stream/server.go:221]: I've moved the "write lock" up to include this.

File: robot/web/stream/server.go
Comment (dgottlieb) [reply robot/web/stream/server.go:260]: Yeah this is just wrong and I'm not doing to touch it. The read-lock I believe is intended for accessing `ss.streams`. And this read-lock will incidentally block out [concurrent callers to](https://github.com/viamrobotics/rdk/pull/3699/files#diff-b664e7f5f24e04be1ed075bf7d0f4209eaf9fc88f34eb7acc987ad1693857223R40-R48) `streamState.Stop`. But it won't block out concurrent callers to `Server.Stop`.

I'm leaving this as-is, but the smell bothers me to. I think we got a little too cute trying to trying to avoid adding a `streamState.Close/Shutdown` that avoids the reference counting.

**Review by nicksanford — APPROVED** [review]

**Pull request merged**
